soc/tenstorrent/tt_blackhole/soc.c

```diff
@@ -16,8 +21,60 @@ static inline void delay_spin(uint32_t count)
 	}
 }
 
+#define PANIC_REASON_MAGIC 0xBADC0DE0
+#define REASON_K_PANIC     0x1
+#define REASON_ARC_RESET   0x2
+
+#define ARC_RESET_ADDR 0x80000000
+
+/*
+ * After panic the arc can't write to scratch registers. Write to ICCM
+ * to indicate we have panicked.
+ */
+#define ARC_PANIC(reason, blink)                                                                   \
+	do {                                                                                       \
+		sys_write32(PANIC_REASON_MAGIC | (reason), 0x0);                                   \
+		sys_write32(blink, 0x4);                                                           \
+	} while (0)
+
+/*
+ * Override the default kernel panic handler. We want to also
+ * dump the BLINK register value.
+ */
+__weak void k_sys_fatal_error_handler(unsigned int reason, const struct arch_esf *esf)
+{
+	ARG_UNUSED(esf);
+
+	ARC_PANIC(REASON_K_PANIC, esf->blink);
```

Comment:
Upstream, they have the "hwinfo" driver API... but that's usually only useful after a reset and assumes that there is non-volatile memory (or that the SoC can properly reset without a power cycle, which isn't the case for us in this product family).

https://docs.zephyrproject.org/apidoc/latest/group__hwinfo__interface.html

What would also be kind of neat is to add a custom Coredump handler, which I think would allow us to add this hook in without a patch. SYS-1169 in the backlog.

IMHO, that still counts as a stability improvement (although it's kind of also a feature too)
https://docs.zephyrproject.org/latest/services/debugging/coredump.html

Could easily do it in the sprint after next as well.

Reply:
Yeah, I don't think hwinfo will get us much here. This handler *probably* won't ever run in a crash, it only would run if the SOC hit the self-reset bit in the ARC MISC CNTL register.

Coredump looks interesting- I think we probably want to get this enabled with the SPI flash longer term. Probably want multiple backends though- if the system is in a bad enough state we really can't trust that SPI programming will work.

I'll also note this change is probably the lowest value (for debugging) in this PR- when the SMC faults, it rarely does so in a way that actually goes through the Zephyr crash handlers. Usually it just crashes in a way where the program counter reads `0x0` and the system is hung until reset.